tfhe/src/integer/server_key/radix_parallel/mod.rs

```diff
@@ -140,45 +142,51 @@ impl ServerKey {
-                let rhs = T::from(carry_blocks);
+                blocks[0] = message_blocks.remove(0);
+                let mut lhs = RadixCiphertext::from(message_blocks);
+                let rhs = RadixCiphertext::from(carry_blocks);
                 self.add_assign_with_carry_parallelized(&mut lhs, &rhs, None);
-                ctxt.blocks_mut()[start_index + 1..].clone_from_slice(lhs.blocks());
+                blocks[1..].clone_from_slice(&lhs.blocks);
             } else {
-                self.propagate_single_carry_parallelized(&mut ctxt.blocks_mut()[start_index..]);
+                self.propagate_single_carry_parallelized(&mut blocks[..]);
             }
         } else {
-            let maybe_highest_degree = ctxt
+            let maybe_highest_degree =
                 // We do not care about degree of 'first' block as it won't receive any carries
-                .blocks()[start_index + 1..]
+                blocks[1..]
                 .iter()
                 .max_by(|block_a, block_b| block_a.degree.get().cmp(&block_b.degree.get()))
                 .map(|block| block.degree.get());
 
+            let mut start_index = 0;
             if maybe_highest_degree.is_some_and(|degree| degree > self.key.max_degree.get()) {
                 // At least one of the blocks than can receive a carry, won't be able too
                 // so we need to do a first 'partial' round
-                let (mut message_blocks, carry_blocks) =
-                    extract_message_and_carry_blocks(&ctxt.blocks()[start_index..]);
-                ctxt.blocks_mut()[start_index..].swap_with_slice(&mut message_blocks);
-                for (block, carry) in ctxt.blocks_mut()[start_index + 1..]
-                    .iter_mut()
-                    .zip(carry_blocks.iter())
-                {
+                let (mut message_blocks, carry_blocks) = extract_message_and_carry_blocks(blocks);
+                blocks[0..].swap_with_slice(&mut message_blocks);
+                for (block, carry) in blocks[1..].iter_mut().zip(carry_blocks.iter()) {
                     self.key.unchecked_add_assign(block, carry);
                 }
                 // We can start propagation one index later as we already did the first block
                 start_index += 1;
             }
 
-            let len = ctxt.blocks().len();
+            let len = blocks.len();
             // If start_index >= len, the range is considered empty
             for i in start_index..len {
-                let _ = self.propagate_parallelized(ctxt, i);
+                let (carry, message) = rayon::join(
+                    || self.key.carry_extract(&blocks[i]),
+                    || self.key.message_extract(&blocks[i]),
+                );
+
+                blocks[i] = message;
+
+                //add the carry to the next block
+                if i < blocks.len() - 1 {
+                    self.key.unchecked_add_assign(&mut blocks[i + 1], &carry);
```

Comment:
we can get out of propagation with non nominal noise ?

Reply:
Yes there should be no problem:
For blocks in range 0..len we extract both the msg and carry, and if the block is not the last, we add the carry to the next block, which will then clean itself when its going to extract its msg and carry

Reply:
indeed, misread the loop